### ISSUE TITLE
TooLongFrameException thrown via docker-java for > 1000 containers running

### ISSUE THREAD

**Issue by fraz3alpha** [jenkinsci/docker-plugin#601]
When I have a large number of containers already running in my target Docker environment the following error is thrown by the Docker plugin (and its dependencies) and containers cannot be created:

```
Jan 10, 2018 9:56:00 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud provision
Asked to provision 1 slave(s) for: jenkins-slave
Jan 10, 2018 9:56:00 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud provision
Will provision 'evarga/jenkins-slave', for label: 'jenkins-slave', in cloud: 'test-swarm'
Jan 10, 2018 9:56:00 AM SEVERE com.github.dockerjava.core.async.ResultCallbackTemplate onError
Error during callback
io.netty.handler.codec.TooLongFrameException: object length exceeds 1048576: 1056644 bytes discarded
	at io.netty.handler.codec.json.JsonObjectDecoder.decode(JsonObjectDecoder.java:106)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)
...
```
This happens when I get to approximately 1100 containers, and can be matched with a call to `curl ${DOCKER_HOST}/containers/json | wc`, so it is probably occuring during the part of the process where the plugin is requesting out what is currently running so as to obey any caps defined.
e.g. 
```
# curl ${DOCKER_HOST}/containers/json | wc
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1092k    0 1092k    0     0  22.2M      0 --:--:-- --:--:-- --:--:-- 22.6M
      1    4243 1118431
```

There is no configurable parameters as far as I can tell to adjust the size of decoding frame

Environment details:

Jenkins version: `2.73.1`
docker-plugin version: `1.1.2`
Docker version `17.09.0-ce` via legacy swarm mode (`swarm:1.2.8`)

My current level of running containers was not a problem at version 0.16.2, it was only with an upgrade to 1.1.2 where this started happening.

In this particular test the 1000+ containers that are running aren't doing anything, they were generated outside of Jenkins and are just sitting there sleeping via a script such as this one:
```
#!/bin/bash
for i in `seq 1 1000`;
do
  echo $i
  docker -H :4000 run -d busybox:latest sleep 18000
done
```
Full output of failing and passing plugin versions:

v0.16.2 is successful:
```
Jan 10, 2018 10:34:52 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud provision
Asked to provision 1 slave(s) for: jenkins-slave
Jan 10, 2018 10:34:52 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud provision
Will provision 'evarga/jenkins-slave', for label: 'jenkins-slave', in cloud: 'test-swarm'
Jan 10, 2018 10:34:52 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud addProvisionedSlave
Provisioning 'evarga/jenkins-slave' number '1' on 'test-swarm'; Total containers: '1201'
Jan 10, 2018 10:34:52 AM INFO hudson.slaves.NodeProvisioner$StandardStrategyImpl apply
Started provisioning Image of evarga/jenkins-slave from test-swarm with 1 executors. Remaining excess workload: -0.122
Jan 10, 2018 10:34:52 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud provisionWithWait
Trying to run container for evarga/jenkins-slave
Jan 10, 2018 10:34:56 AM INFO com.nirima.jenkins.plugins.docker.utils.PortUtils$ConnectionCheckSSH execute
SSH port is open on 10.127.159.67:40026
Jan 10, 2018 10:34:56 AM INFO com.nirima.jenkins.plugins.docker.launcher.DockerComputerSSHLauncher getSSHLauncher
Creating slave SSH launcher for /10.127.159.67:40026
Jan 10, 2018 10:35:02 AM INFO hudson.slaves.NodeProvisioner$2 run
Image of evarga/jenkins-slave provisioning successfully completed. We have now 3 computer(s)
```

v1.1.2 fails:
```
Jan 10, 2018 10:43:49 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud provision
Asked to provision 1 slave(s) for: jenkins-slave
Jan 10, 2018 10:43:49 AM INFO com.nirima.jenkins.plugins.docker.DockerCloud provision
Will provision 'evarga/jenkins-slave', for label: 'jenkins-slave', in cloud: 'test-swarm'
Jan 10, 2018 10:43:49 AM SEVERE com.github.dockerjava.core.async.ResultCallbackTemplate onError
Error during callback
io.netty.handler.codec.TooLongFrameException: object length exceeds 1048576: 1049308 bytes discarded
	at io.netty.handler.codec.json.JsonObjectDecoder.decode(JsonObjectDecoder.java:106)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:72)
	at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:33)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:297)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:413)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)

Jan 10, 2018 10:43:49 AM WARNING com.nirima.jenkins.plugins.docker.DockerCloud provision
Bad template 'evarga/jenkins-slave' in cloud 'test-swarm': 'object length exceeds 1048576: 1049308 bytes discarded'. Trying next template...
io.netty.handler.codec.TooLongFrameException: object length exceeds 1048576: 1049308 bytes discarded
	at io.netty.handler.codec.json.JsonObjectDecoder.decode(JsonObjectDecoder.java:106)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:72)
	at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:33)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:297)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:413)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by fraz3alpha** [comment]
Here is some further information regarding the different parts in this, and I believe that ultimately it's probably not this plugin's fault as none of the dependencies provide a way to alter the behaviour to use a bigger buffer size, but it does mean it is broken.

There is nothing we can configure in the [docker-java-api-plugin](https://github.com/jenkinsci/docker-java-api-plugin) that only bundles the library

The netty library is where the [JsonObjectDecoder() method](https://github.com/netty/netty/blob/4.1/codec/src/main/java/io/netty/handler/codec/json/JsonObjectDecoder.java#L65) is defined (which throws this error)

The docker-java project has [several](https://github.com/docker-java/docker-java/blob/master/src/main/java/com/github/dockerjava/netty/InvocationBuilder.java#L203) [different](https://github.com/docker-java/docker-java/blob/master/src/main/java/com/github/dockerjava/netty/InvocationBuilder.java#L352) [instances](https://github.com/docker-java/docker-java/blob/master/src/main/java/com/github/dockerjava/netty/InvocationBuilder.java#L458) of calls to this method, but no options are passed to the constructor, so the default of [`1024 * 1024`](https://github.com/netty/netty/blob/4.1/codec/src/main/java/io/netty/handler/codec/json/JsonObjectDecoder.java#L56) is used for the frame size

**Comment by alexeygrigorovaurea** [comment]
I was facing same issue even with 100 containers, and had to  build custom docker-java fork  with non-default size.

**Comment by ndeloof** [comment]
0.16.x used the Jersey docker-java implementation, so wasn't affected the same way by such a large load. Not sure there's anything we can do here but wait for docker-java to provide a fix / parameter, or for me to complete migration to [jocker](https://github.com/dockins/jocker) client library.

**Comment by fraz3alpha** [comment]
I don't think docker-java know this is a problem, so I'll raise an issue on their github with the appropriate parts of the above information.

**Comment by pjdarton** [comment]
I'm seeing much the same issues where I work, as we've got a shared docker swarm that can get quite busy at times, so all the details of "all docker containers" is quite a big download and trips over the 1meg threshold that docker-java defaults to when using Netty.

I think I've come up with a fix in https://github.com/jenkinsci/docker-plugin/pull/616 which is to ensure that, when asking docker to list the containers, we ask it to only list the ones _we_ started (with a further workaround available that, if you still hit this limit, don't specify a per-cloud instance cap, and if you still hit this limit due to having thousands of containers of a single template, stop having a per-template instance cap too).

Anyone here who's interested in this issue is welcome to try out https://ci.jenkins.io/job/Plugins/job/docker-plugin/job/PR-616/1/artifact/target/docker-plugin.hpi and see if solves the issue.

FYI I don't think docker-java faults that only affect Netty and not Jersey are considered as a high priority - (years ago) the primary owner/maintainer of docker-java was the original author of (this) docker-plugin code, but then there was a disagreement and he forked the code to create yet-another-docker-plugin; yet-another uses Jersey, so it isn't affected by these issues, so there's little incentive to fix Netty-only issues.

**Comment by pjdarton** [comment]
#616 was merged in to version 1.1.3.
This restricts the scope of the "list containers in host" calls to only include those containers started by the plugin for this Jenkins server, so that should mean that we can now cope with a busy swarm (as long as we're not responsible for 1000 containers ourselves).

I'd suggest that folks interested in this issue try version 1.1.3 and see if that fixes it for them.

**Comment by pjdarton** [comment]
Closing: I believe this is now fixed.
Feel free to re-open if this (exact-same-issue) is still an issue.
